### PR TITLE
Explicitly ignore absolute paths for gradle `Test` task inputs

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

It has been noted that instead of using relative paths, we can use absolute paths for test inputs as long as we mark the necessary property as ignored.
By doing so, we may take further advantage of remote build caches.

Testing has been done by decreasing the heap to simulate failure and checking if the artifacts are uploaded in the correct location.

- https://github.com/line/armeria/actions/runs/5712843311?pr=5082
- https://github.com/line/armeria/pull/5082/commits/eeae36183bedf3fd44ba1967bfd56c156020ea97

Modifications:

- Introduce a `cliArgProvider` which takes a closure and returns cli arguments.
- Migrate `blockhound`, heapdump path related fields to use `cliArgProvider`.

Result:

- More explicit/clear usage of build directory property in gradle scripts.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -119,6 +118,25 @@ allprojects {
     }
 }
 
+/**
+ * An argument provider which avoids task input inconsistency when builds are run
+ * from different locations. By handling this, we can take advantage of build caches
+ * when run from different locations. (i.e. remote build caches)
+ */
+class TestBuildDirArgumentsProvider implements CommandLineArgumentProvider {
```
Comment (trustin) [inline build.gradle:126]: Can we move this to `/buildSrc` so we can reuse it in other places in our build?

File: build.gradle [hunk 2]
```diff
@@ -63,13 +63,19 @@ allprojects {
     }
 
     tasks.withType(Test) {
+        jvmArgumentProviders.add(project.ext.cliArgProvider({
+            [
+                    "-Dcom.linecorp.armeria.blockhound.reportFile=${project.buildDir}/blockhound.log",
+                    "-XX:HeapDumpPath=${project.buildDir}",
```
Comment (ribafish) [inline build.gradle:69]: My suggestion would be to not overgeneralize the `cliArgProvider`, but instead create one specific for tests, set the `blockhound` as an output file with relative path sensitivity and the `HeapDumpPath` as `@LocalState`, which will also get removed if the task is `FROM-CACHE`.

**Review by trustin — APPROVED** [review]
:+1: Thanks @ribafish for the feedback!

**Review by ikhoon — APPROVED** [review]
Thanks, @jrhee17!

**Pull request merged**